content/articles/lru-cache-implementation-in-javascript/index.md

```diff
@@ -0,0 +1,192 @@
+
+Below are the steps we will use to implement the LRU Cache class
+
+1. We first initialize the LRU cache with a positive capacity.
+Copy the below code to implement the constructor on line number 4. 
+
+```javascript
+var LRUCache = function (capacity) {
+    this.capacity = capacity;
+    this.map = new Map(); //this stores the entire array
+
+    //this is boundaries for double Linked List
+    this.head = {};
+    this.tail = {};
+    this.head.next = this.tail; //initialize your double Linked List
+    this.tail.prev = this.head;
+};
+```
+2.  Get operation. This will return the value of the Key if it exists. Else, it returns -1.
+The code below will implement the get function on line number 12.
+
+```javascript
+LRUCache.prototype.get = function (key) {
+    if (this.map.has(key)) {
+        //remove elem from current position
+        let c = this.map.get(key);
+        c.prev.next = c.next;
+        c.next.prev = c.prev;
+
+        this.tail.prev.next = c; //insert it after last element. Element before tail
+        c.prev = this.tail.prev; //update c.prev and next pointer
+        c.next = this.tail;
+        this.tail.prev = c; //update last element as tail
+        return c.value;
+    } else {
+        return -1; //element does not exist
+    }
+
+};
+```
+
+3.  Put operation. The put operation will update the value of the key if it is found. If found, add the `Key` and the value pair to the cache. If the number of keys has exceeded the initialized capacity of the cache, evict the least recently accessed item. 
+The code below will implement the put function on line number 21
+
+```javascript
+LRUCache.prototype.put = function (key, value) {
+
+    if (this.get(key) !== -1) { // if key does not exist, update last element value
+        this.tail.prev.value = value;
+    } else {
+        //need to check if map size is at capacity
+        if (this.map.size === this.capacity) {
+            //delete item both from map and DLL
+            this.map.delete(this.head.next.key); //delete first element of list
+            this.head.next = this.head.next.next; //update first element as next element
+            this.head.next.prev = this.head;
+        }
+
+        let newNode = {
+            value,
+            key
+        }; //each node is a hashtable that stores key and value
+
+        //When adding a new node, we need to update both map and DLL
+        this.map.set(key, newNode); //add current node to map
+        this.tail.prev.next = newNode; //add node to end of the list
+        newNode.prev = this.tail.prev; //update prev and next pointers of newNode
+        newNode.next = this.tail;
+        this.tail.prev = newNode; //update last element
+    }
+
+};
+```
+Once done, click on the run code button. It is located on the bottom right. Below is the explanation of the expected output.
+
+Example
+
+#### Input
+```javascript
+["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
+

```

Comment:
Everything in this section and the ones below it are copied verbatim from the LeetCode challenge. Please remove them and use your own words to explain and guide the reader on how to run the code.

**Use step by step instructions, explaining what is happening and what is expected in each stage**